packages/menu/src/useMenu.ts

```diff
@@ -0,0 +1,654 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import {
+  RefObject,
+  createRef,
+  useCallback,
+  useEffect,
+  useMemo,
+  useReducer,
+  useRef,
+  useState
+} from 'react';
+import { useSelection } from '@zendeskgarden/container-selection';
+import {
+  KEYS,
+  getControlledValue,
+  composeEventHandlers,
+  useId
+} from '@zendeskgarden/container-utilities';
+import {
+  isItemGroup,
+  isValidItem,
+  StateChangeTypes,
+  stateReducer,
+  toMenuItemKeyDownType
+} from './utils';
+import {
+  MenuItem,
+  IMenuItemBase,
+  IUseMenuProps,
+  IUseMenuReturnValue,
+  ItemValue,
+  MenuItemGroup,
+  ISelectedItem
+} from './types';
+
+export const useMenu = <T extends HTMLElement = HTMLElement, L extends HTMLElement = HTMLElement>({
+  items: rawItems,
+  idPrefix,
+  environment,
+  menuRef,
+  triggerRef,
+  onChange = () => undefined,
+  isExpanded,
+  initialExpanded = false,
+  selectedItems,
+  focusedValue,
+  initialFocusedValue,
+  defaultFocusedValue
+}: IUseMenuProps<T, L>): IUseMenuReturnValue<T, L> => {
+  const prefix = `${useId(idPrefix)}-`;
+  const triggerId = `${prefix}menu-trigger`;
+  const isExpandedControlled = isExpanded !== undefined;
+  const isSelectionValueControlled = selectedItems !== undefined;
+  const isFocusedValueControlled = focusedValue !== undefined;
+
+  const menuItems = useMemo(
+    () =>
+      rawItems.reduce((items, item: MenuItem) => {
+        if (isItemGroup(item)) {
+          items.push(...((item as MenuItemGroup).items.filter(isValidItem) as IMenuItemBase[]));
+        } else if (isValidItem(item)) {
+          items.push(item as IMenuItemBase);
+        }
+
+        return items;
+      }, [] as IMenuItemBase[]),
+    [rawItems]
+  );
+
+  const values = useMemo(() => menuItems.map(item => item.value), [menuItems]);
+
+  const itemRefs = useMemo(
+    () =>
+      values.reduce((acc: Record<ItemValue, RefObject<any | null>>, v) => {
+        acc[v] = createRef();
+
+        return acc;
+      }, {}),
+    [values]
+  );
+
+  /**
+   * Validate controlled params
+   */
+
+  if (selectedItems !== undefined && selectedItems !== null) {
+    if (!Array.isArray(selectedItems)) {
```

Comment:
Do you need the additional guard `if (selectedItems !== undefined && selectedItems !== null)` when using `Array.isArray`? Being that we only want to check if it's an array, both `undefined` and `null` will return `false` when evaluated with `Array.isArray`.

Reply:
Another thought, could lean on TypeScript to guard against unexpected inputs in this case?

Reply:
Very true re: `Array.isArray` - I'll fix that up.

For TS, That's also a fair point. Either way we'll throw an error because non-array, non-nullish `selectedItems` will break controlled selection. This just catches it early I suppose. I see a similar thing going on in `useCombobox` but its arg can be also be a string. 🤔

Reply:
I removed the whole block since the typing is singular. 👍🏻